### PR TITLE
Move settings to options

### PR DESCRIPTION
**Pull request by golopot**
Fixes part of #216.

This pr moves some settings to rule options. When a removed setting is set, a warning or error will be reported.

Does not move settings for `check-examples` because it is too much work.

BREAKING CHANGE: The following settings has been removed and became options
    under relevant rules.
    * `settings.jsdoc.allowEmptyNamepaths` -> option in `valid-types`
    * `settings.jsdoc.checkSeesForNamepaths` -> option in `valid-types`
    * `settings.jsdoc.exemptEmptyFunctions` -> option in `require-jsdoc`
    * `settings.jsdoc.forceRequireReturn` -> option in `require-returns`
    * `settings.jsdoc.avoidExampleOnConstructors` -> option in `require-example`

### CODE REVIEW

File: src/rules/requireExample.js [hunk 1]
```diff
@@ -1,23 +1,30 @@
...
 }) => {
+  warnRemovedSettings(context, 'require-example');
+
   if (utils.avoidDocs()) {
     return;
   }
 
+  const options = context.options[0] || {
+    avoidExampleOnConstructors: false
+  };
```
Comment (brettz9) [inline src/rules/requireExample.js]: I might suggest something like this here instead

```js
  const {avoidExampleOnConstructors = false} = context.options[0] || {};
```

Besides being potentially shorter, this allows for the `avoidExampleOnConstructors` default to be applied in the future if other options get added, e.g., if the user had their options object as `{someOtherOption: true}`, the code would only need to be changed to this, while keeping the default of `avoidExampleOnConstructors` as `false`.

```js
  const {avoidExampleOnConstructors = false, someOtherOption} = context.options[0] || {};
```

Same with the other block like this below.

**Comment by brettz9** [comment]
Besides the suggestion for a more future-proof structure for option defaults, all looks good to me!

File: src/rules/requireReturns.js [hunk 2]
```diff
@@ -43,16 +44,20 @@ const canSkip = (utils) => {
...
   // in case the @returns comment is optional or undefined.
   if (canSkip(utils)) {
     return;
   }
 
-  const options = context.options[0] || {};
+  const options = context.options[0] || {
+    forceRequireReturn: false,
+    forceReturnsWithAsync: false
+  };
```
Comment (brettz9) [inline src/rules/requireReturns.js]: Here too.

File: src/rules/validTypes.js [hunk 3]
```diff
@@ -1,13 +1,22 @@
...
+  utils,
+  context
 }) => {
+  warnRemovedSettings(context, 'valid-types');
+
+  const options = context.options[0] || {
+    allowEmptyNamepaths: true,
+    checkSeesForNamepaths: false
+  };
+
```
Comment (brettz9) [inline src/rules/validTypes.js]: and here...

File: .README/rules/require-example.md [hunk 4]
```diff
@@ -7,11 +7,15 @@ Requires that all functions have examples.
...
 - `exemptedBy` - Array of tags (e.g., `['type']`) whose presence on the document
   block avoids the need for an `@example`. Defaults to an empty array.
 
+- `avoidExampleOnConstructors` (default: false) - Set to `true` to avoid the
+  need for an example on a constructor (whether indicated as such by a
+  jsdoc tag or by being within an ES6 `class`).
+
 |||
 |---|---|
 |Context|`ArrowFunctionExpression`, `FunctionDeclaration`, `FunctionExpression`|
```
Comment (brettz9) [inline .README/rules/require-example.md]: Oh, and the option additions need to be added to these documentation tables...

File: .README/rules/require-jsdoc.md [hunk 5]
```diff
@@ -36,6 +36,10 @@ Accepts one optional options object with the following optional keys.
   AST contexts where you wish the rule to be applied (e.g., `Property` for
   properties). Defaults to an empty array.
 
+- `exemptEmptyFunctions` (default: false) - When `true`, the rule will not report
+  missing jsdoc blocks above functions/methods with no parameters or return values
+  (intended where variable names are sufficient for themselves as documentation).
+
 |||
 |---|---|
 |Context|`ArrowFunctionExpression`, `ClassDeclaration`, `ClassExpression`, `FunctionDeclaration`, `FunctionExpression`|
```
Comment (brettz9) [inline .README/rules/require-jsdoc.md]: and within this table...

File: .README/rules/require-returns.md [hunk 6]
```diff
@@ -6,6 +6,10 @@ Requires returns are documented.
 
 - `exemptedBy` - Array of tags (e.g., `['type']`) whose presence on the document
     block avoids the need for a `@returns`. Defaults to an empty array.
+- `forceRequireReturn` - Set to `true` to always insist on
+  `@returns` documentation regardless of implicit or explicit `return`'s
+  in the function. May be desired to flag that a project is aware of an
+  `undefined`/`void` return. Defaults to `false`.
 - `forceReturnsWithAsync` - By default `async` functions that do not explicitly return a value pass this rule. You can force all `async` functions to require return statements by setting `forceReturnsWithAsync` to `true` on the options object. This may be useful as an `async` function will always return a `Promise`, even if the `Promise` returns void. Defaults to `false`.
 
 ```js
```
Comment (brettz9) [inline .README/rules/require-returns.md]: and in the table in these docs...

File: .README/rules/valid-types.md [hunk 7]
```diff
@@ -32,6 +32,21 @@ Also impacts behaviors on namepath (or event)-defining and pointing tags:
...
+  may also allow names to be defined in another manner elsewhere in
+  the block)
+- `checkSeesForNamepaths` (default: false) - Set this to `true` to insist
+  that `@see` only use name paths (the tag is normally permitted to
+  allow other text)
+
+
 |||
 |---|---|
 |Context|everywhere|
```
Comment (brettz9) [inline .README/rules/valid-types.md]: and in this table...

**Comment by golopot** [comment]
ready for review

**Pull request merged**